src/viam/trajex/totg/trajectory.cpp

```diff
@@ -488,79 +572,100 @@ enum class integration_event : std::uint8_t {
+                before = mid;
+            }
         }
 
-        const double curve_slope = compute_velocity_limit_derivative(q_prime, q_double_prime, opt.max_velocity, opt.epsilon);
+        // Store the refined continuous switching point
+        search_cursor.seek(after);
+        const auto q_prime = search_cursor.tangent();
+        const auto q_double_prime = search_cursor.curvature();
+        const auto [s_dot_max_acc, s_dot_max_vel] =
+            compute_velocity_limits(q_prime, q_double_prime, opt.max_velocity, opt.max_acceleration, opt.epsilon);
 
-        const auto [s_ddot_min, s_ddot_max] =
-            compute_acceleration_bounds(q_prime, q_double_prime, s_dot_max_vel, opt.max_acceleration, opt.epsilon);
+        // TODO: Implement Addendum E validation - reject switching points where s_dot_max_vel > s_dot_max_acc.
+        // Per the reference implementation (Trajectory.cpp:129-133), if the velocity switching point has
+        // s_dot > s_dot_max_acc, backward integration will immediately hit the acceleration limit curve,
+        // causing the algorithm to fail. We should continue searching forward until finding a switching
+        // point where s_dot_max_vel <= s_dot_max_acc + epsilon.
 
-        const double trajectory_slope = s_ddot_min / s_dot_max_vel;
+        continuous_switching_point = trajectory::phase_point{.s = after, .s_dot = s_dot_max_vel};
+    }
 
-        if (trajectory_slope <= curve_slope + opt.epsilon) {
-            // Midpoint satisfies escape condition - narrow to [before, mid]
-            after = mid;
+    // Phase 4: Return whichever switching point comes first.
+    // If both exist, return the one with smaller s (earlier along the path).
+    // If only one exists, return that one.
+    // If neither exists, return end of path.
+    if (discontinuous_switching_point.has_value() && continuous_switching_point.has_value()) {
+        // Both types found - return whichever is earlier
+        if (discontinuous_switching_point->s < continuous_switching_point->s) {
+            return *discontinuous_switching_point;
+        } else if (continuous_switching_point->s < discontinuous_switching_point->s) {
+            return *continuous_switching_point;
         } else {
-            // Midpoint doesn't satisfy escape - narrow to [mid, after]
-            before = mid;
+            // Same location (within epsilon) - use whichever has lower velocity (more conservative)
+            return (discontinuous_switching_point->s_dot < continuous_switching_point->s_dot) ? *discontinuous_switching_point
+                                                                                              : *continuous_switching_point;
         }
+    } else if (discontinuous_switching_point.has_value() && !continuous_switching_point.has_value()) {
+        return *discontinuous_switching_point;
+    } else if (continuous_switching_point.has_value() && !discontinuous_switching_point.has_value()) {
```

Comment:
These checks are partially redundant. You know you don't have both, so you can just check each case on its own.